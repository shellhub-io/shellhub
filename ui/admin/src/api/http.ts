import { AxiosInstance } from "axios";
import { setupInterceptorsTo } from "./interceptors";
import * as axiosTs from "./client";
import { Configuration } from "./client";
import { BaseAPI } from "./client/base";

const configuration = new Configuration();
configuration.basePath = `${window.location.protocol}//${window.location.host}`;
configuration.accessToken = localStorage.getItem("cloud_token") || "";

// eslint-disable-next-line import/no-mutable-exports
let adminApi = new axiosTs.AdminApi(configuration);
// eslint-disable-next-line import/no-mutable-exports
let cloudApi = new axiosTs.CloudApi(configuration);

export const createNewClient = () => {
  const newConfiguration = new Configuration();
  newConfiguration.basePath = `${window.location.protocol}//${window.location.host}`;
  newConfiguration.accessToken = localStorage.getItem("cloud_token") || "";
  adminApi = new axiosTs.AdminApi(newConfiguration);
  cloudApi = new axiosTs.CloudApi(newConfiguration);
  return { adminApi, cloudApi };
};

declare module "./client/base" {
  interface BaseAPI {
    getAxios(): AxiosInstance;
    getConfiguration(): Configuration | undefined;
    setConfiguration(configuration: Configuration): void;
  }
}

/** Returns the axios instance */
BaseAPI.prototype.getAxios = function getAxios(this: BaseAPI): AxiosInstance {
  return this.axios;
};

/** Returns the configuration */
// eslint-disable-next-line vue/max-len
BaseAPI.prototype.getConfiguration = function getConfiguration(this: BaseAPI): Configuration | undefined {
  return this.configuration;
};

/** Sets the configuration */
// eslint-disable-next-line vue/max-len
BaseAPI.prototype.setConfiguration = function setConfiguration(this: BaseAPI, configuration: Configuration): void {
  this.configuration = configuration;
};

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
setupInterceptorsTo(adminApi.axios);

export { adminApi, cloudApi };
