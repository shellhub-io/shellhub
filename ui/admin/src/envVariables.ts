/* eslint-disable import/prefer-default-export */
export const envVariables = {
  announcementsEnable: process.env.SHELLHUB_ANNOUNCEMENTS === "true",
  tinyMceKey: process.env.TINYMCE_API_KEY,
};
