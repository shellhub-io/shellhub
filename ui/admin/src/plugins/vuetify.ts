// Styles
// import "@mdi/font/css/materialdesignicons.css";
import "vuetify/styles";

// Vuetify
import { createVuetify } from "vuetify";

export default createVuetify({
  theme: {
    defaultTheme: "dark",
  },
});
