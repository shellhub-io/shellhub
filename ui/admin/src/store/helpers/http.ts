import Axios from "axios";
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore //* use while route mapping is not in TypeScript
export default () => {
  const axios = Axios.create({
    baseURL: `${window.location.protocol}//${window.location.host}/admin`,
    headers: {
      Authorization: `Bearer ${localStorage.getItem("cloud_token")}`,
    },
  });

  return axios;
};
