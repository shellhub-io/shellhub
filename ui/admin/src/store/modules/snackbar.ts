import { Module } from "vuex";
import { State } from "./../index";

export interface snackbarMessageAndContentType {
  typeMessage: string;
  typeContent: string;
}

export interface SnackbarState {
  snackbarError: boolean;
  snackbarSuccess: boolean;
  snackbarMessageAndContentType: snackbarMessageAndContentType;
  snackbarCopy: boolean;
}

export const snackbar: Module<SnackbarState, State> = {
  namespaced: true,

  state: {
    snackbarError: false,
    snackbarSuccess: false,
    snackbarMessageAndContentType: { typeMessage: "", typeContent: "" },
    snackbarCopy: false,
  },

  getters: {
    snackbarSuccess: (state) => state.snackbarSuccess,
    snackbarError: (state) => state.snackbarError,
    snackbarMessageAndContentType: (state) => state.snackbarMessageAndContentType,
    snackbarCopy: (state) => state.snackbarCopy,
  },

  mutations: {
    setSnackbarSuccessAction: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarSuccess = true;
    },

    setSnackbarNoContent: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarSuccess = true;
    },

    setSnackbarSuccessDefault: (state) => {
      state.snackbarMessageAndContentType = { typeMessage: "default", typeContent: "" };
      state.snackbarSuccess = true;
    },

    unsetSnackbarSuccess: (state) => {
      state.snackbarSuccess = false;
    },

    setSnackbarErrorLoadingOrAction: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarError = true;
    },

    setSnackbarErrorCustom: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarError = true;
    },

    setSnackbarErrorDefault: (state) => {
      state.snackbarMessageAndContentType = { typeMessage: "default", typeContent: "" };
      state.snackbarError = true;
    },

    setSnackbarErrorLicense: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarError = true;
    },

    unsetSnackbarError: (state) => {
      state.snackbarError = false;
    },

    setSnackbarCopy: (state, value) => {
      state.snackbarMessageAndContentType = { typeMessage: "", typeContent: value };
      state.snackbarCopy = true;
    },

    unsetSnackbarCopy: (state) => {
      state.snackbarCopy = false;
    },
  },

  actions: {
    showSnackbarSuccessAction: ({ commit }, value) => {
      const data = { typeMessage: "action", typeContent: value };
      commit("setSnackbarSuccessAction", data);
    },

    showSnackbarSuccessDefault: ({ commit }) => {
      commit("setSnackbarSuccessDefault");
    },

    unsetShowStatusSnackbarSuccess: ({ commit }) => {
      commit("unsetSnackbarSuccess");
    },

    showSnackbarErrorLoading: ({ commit }, value) => {
      const data = { typeMessage: "loading", typeContent: value };
      commit("setSnackbarErrorLoadingOrAction", data);
    },

    showSnackbarNoContent: ({ commit }) => {
      const data = { typeMessage: "no-content", typeContent: "" };
      commit("setSnackbarNoContent", data);
    },

    showSnackbarErrorAction: ({ commit }, value) => {
      const data = { typeMessage: "action", typeContent: value };
      commit("setSnackbarErrorLoadingOrAction", data);
    },

    showSnackbarErrorCustom: ({ commit }, value) => {
      const data = { typeMessage: "custom", typeContent: value };
      commit("setSnackbarErrorCustom", data);
    },

    showSnackbarErrorLicense: ({ commit }, value) => {
      const data = { typeMessage: "licenseRequired", typeContent: value };
      commit("setSnackbarErrorLicense", data);
    },

    showSnackbarErrorDefault: ({ commit }) => {
      commit("setSnackbarErrorDefault");
    },

    unsetShowStatusSnackbarError: ({ commit }) => {
      commit("unsetSnackbarError");
    },

    showSnackbarCopy: ({ commit }, value) => {
      commit("setSnackbarCopy", value);
    },

    unsetShowStatusSnackbarCopy: ({ commit }) => {
      commit("unsetSnackbarCopy");
    },
  },
};
