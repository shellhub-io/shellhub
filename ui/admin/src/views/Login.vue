<template>
  <v-app>
    <v-main>
      <v-container class="full-height d-flex justify-center align-center" fluid>
        <v-row align="center" justify="center">
          <v-col cols="12" sm="8" md="4">
            <v-card theme="dark" class="pa-6" rounded="lg">
              <v-card-title class="d-flex justify-center align-center mt-4">
                <v-img
                  :src="Logo"
                  max-width="220"
                  alt="logo do ShellHub, uma nuvem de com a escrita ShellHub Admin ao lado"
                />
                <span class="mt-6 text-overline">Admin</span>
              </v-card-title>
              <v-card-item>
                <SnackbarComponent />
                <form @submit.prevent="login">
                  <v-container>
                    <v-text-field
                      color="primary"
                      prepend-icon="mdi-account"
                      v-model="username"
                      :error-messages="usernameError"
                      required
                      label="Username"
                      variant="underlined"
                      data-test="username-text"
                    />

                    <v-text-field
                      color="primary"
                      prepend-icon="mdi-lock"
                      :append-inner-icon="showPassword ? 'mdi-eye' : 'mdi-eye-off'"
                      v-model="password"
                      :error-messages="passwordError"
                      label="Password"
                      required
                      variant="underlined"
                      data-test="password-text"
                      :type="showPassword ? 'text' : 'password'"
                      @click:append-inner="showPassword = !showPassword"
                    />
                    <v-card-actions class="justify-center">
                      <v-btn
                        type="submit"
                        data-test="login-btn"
                        color="primary"
                        variant="tonal"
                        block
                        @click="login"
                      >
                        LOGIN
                      </v-btn>
                    </v-card-actions>
                  </v-container>
                </form>
              </v-card-item>
            </v-card>
          </v-col>
        </v-row>
      </v-container>
    </v-main>
  </v-app>
</template>

<script setup lang="ts">
import { ref } from "vue";
import { useField } from "vee-validate";
import * as yup from "yup";
import { useRoute, useRouter } from "vue-router";
import useAuthStore from "@admin/store/modules/auth";
import useSnackbar from "@/helpers/snackbar";
import Logo from "../assets/logo-inverted.png";
import { createNewClient } from "../api/http";

const showPassword = ref(false);
const snackbar = useSnackbar();
const authStore = useAuthStore();
const route = useRoute();
const router = useRouter();

const { value: username, errorMessage: usernameError } = useField<string | undefined>(
  "name",
  yup.string().required(),
  { initialValue: "" },
);

const { value: password, errorMessage: passwordError } = useField<string | undefined>(
  "password",
  yup.string().required(),
  { initialValue: "" },
);

const hasErrors = () => {
  if (usernameError.value || passwordError.value) {
    return true;
  }

  return false;
};

const login = async () => {
  if (!hasErrors() && username.value && password.value) {
    try {
      await authStore.login({
        username: username.value,
        password: password.value,
      });
      createNewClient();
      if (route.query.redirect) {
        router.push(`${route.query.redirect}`);
      } else {
        router.push("/");
      }
    } catch (error) {
      snackbar.showError("Failed to log in. Please check your credentials and try again.");
    }
  } else {
    snackbar.showError("Failed to log in. Please check your credentials and try again.");
  }
};

defineExpose({ username, password, usernameError, passwordError });
</script>

<style>
.full-height {
  height: 100vh;
}

.v-field__append-inner {
  cursor: pointer;
}
</style>
