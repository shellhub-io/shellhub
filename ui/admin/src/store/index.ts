import { InjectionKey } from "vue";
import { createStore, Store, useStore as vuexUseStore } from "vuex";
import { stats, StatsState } from "./modules/stats";

import { UsersState, users } from "./modules/users";
import { AuthState, auth } from "./modules/auth";
import { DevicesState, devices } from "./modules/devices";
import { firewallRules, FirewallRulesState } from "./modules/firewall_rules";
import { namespaces, NamespacesState } from "./modules/namespaces";
import { layout, LayoutState } from "./modules/layout";
import { license, LicenseState } from "./modules/license";
import { publicKeys, PublicKeysState } from "./modules/public_keys";
import { sessions, SessionsState } from "./modules/sessions";
import { snackbar, SnackbarState } from "./modules/snackbar";
import { spinner, SpinnerState } from "./modules/spinner";
import { announcement, AnnouncementState } from "./modules/announcement";
import { instance, InstanceState } from "./modules/instance";

export interface State {
  user: UsersState;
  auth: AuthState;
  devices: DevicesState;
  firewallRules: FirewallRulesState;
  layout: LayoutState;
  license: LicenseState;
  namespaces: NamespacesState;
  publicKeys: PublicKeysState;
  sessions: SessionsState;
  snackbar: SnackbarState;
  spinner: SpinnerState;
  stats: StatsState;
  announcement: AnnouncementState;
  instance: InstanceState;
}

// eslint-disable-next-line symbol-description
export const key: InjectionKey<Store<State>> = Symbol();

export const store = createStore<State>({
  modules: {
    users,
    auth,
    devices,
    firewallRules,
    layout,
    license,
    namespaces,
    publicKeys,
    sessions,
    snackbar,
    spinner,
    stats,
    announcement,
    instance,
  },
});

export function useStore(): Store<State> {
  return vuexUseStore(key);
}
