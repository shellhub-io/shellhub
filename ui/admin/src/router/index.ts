import { createRouter, createWebHistory, NavigationGuardNext, RouteLocationNormalized } from "vue-router";
import Login from "@admin/views/Login.vue";
import Dashboard from "@admin/views/Dashboard.vue";
import Users from "@admin/views/Users.vue";
import SettingsLicense from "@admin/components/Settings/SettingsLicense.vue";
import SettingsAuthentication from "@admin/components/Settings/SettingsAuthentication.vue";
import Namespaces from "@admin/views/Namespaces.vue";
import Settings from "@admin/views/Settings.vue";
import { computed } from "vue";
import useLicenseStore from "@admin/store/modules/license";
import useLayoutStore from "@admin/store/modules/layout";
import useAuthStore from "@admin/store/modules/auth";
import { plugin as snackbar } from "@/plugins/snackbar"; // using direct plugin because inject() doesn't work outside components

const routes = [
  {
    path: "/login",
    name: "login",
    component: Login,
    meta: {
      layout: "SimpleLayout",
      requiresAuth: false,
    },
  },
  {
    path: "/",
    name: "dashboard",
    component: Dashboard,
  },
  {
    path: "/users",
    name: "users",
    component: Users,
  },
  {
    path: "/user/:id",
    name: "userDetails",
    component: () => import("@admin/views/UserDetails.vue"),
  },
  {
    path: "/devices",
    name: "devices",
    component: () => import("@admin/views/Device.vue"),
    redirect: {
      name: "listDevices",
    },
    children: [
      {
        path: "",
        name: "listDevices",
        component: () => import("@admin/views/Device.vue"),
      },
    ],
  },
  {
    path: "/device/:id",
    name: "deviceDetails",
    component: () => import("@admin/views/DeviceDetails.vue"),
  },
  {
    path: "/sessions",
    name: "sessions",
    component: () => import("@admin/views/Sessions.vue"),
  },
  {
    path: "/session/:id",
    name: "sessionDetails",
    component: () => import("@admin/views/SessionDetails.vue"),
  },
  {
    path: "/settings",
    name: "Settings",
    component: Settings,
    redirect: { name: "SettingProfile" },
    children: [
      {
        path: "authentication",
        name: "SettingAuthentication",
        component: SettingsAuthentication,
      },
      {
        path: "license",
        name: "SettingLicense",
        component: SettingsLicense,
      },
    ],
  },
  {
    path: "/firewall-rules",
    name: "firewall-rules",
    component: () => import("@admin/views/FirewallRules.vue"),
  },
  {
    path: "/firewall-rules/:id",
    name: "firewallRulesDetails",
    component: () => import("@admin/views/FirewallRulesDetails.vue"),
  },
  {
    path: "/namespaces",
    name: "namespaces",
    component: Namespaces,
  },
  {
    path: "/namespace/:id",
    name: "namespaceDetails",
    component: () => import("@admin/views/NamespaceDetails.vue"),
  },
  {
    path: "/announcements",
    name: "announcements",
    component: () => import("@admin/views/Announcements.vue"),
  },
  {
    path: "/announcement/:uuid",
    name: "announcementDetails",
    component: () => import("@admin/views/AnnouncementDetails.vue"),
  },
  {
    path: "/new-announcement",
    name: "new-announcement",
    component: () => import("@admin/views/NewAnnouncement.vue"),
  },
];

const router = createRouter({
  history: createWebHistory("/admin/"),
  routes,
});

router.beforeEach(
  async (to: RouteLocationNormalized, from: RouteLocationNormalized, next: NavigationGuardNext) => {
    const licenseStore = useLicenseStore();
    const layoutStore = useLayoutStore();
    const authStore = useAuthStore();

    const requiresAuth = to.meta.requiresAuth ?? true;

    const layout = to.meta.layout || "AppLayout";
    layoutStore.setLayout(layout as string);

    if (!authStore.isLoggedIn && requiresAuth) {
      return next({
        name: "login",
        query: { redirect: to.fullPath },
      });
    }

    if (authStore.isLoggedIn && !to.meta.requiresAuth) {
      const license = computed(() => licenseStore.getLicense);

      try {
        await licenseStore.get();

        if (license.value.expired && to.name !== "SettingLicense") {
          snackbar.showError("Your license has expired. Please update it and try again.");
          return next({ name: "SettingLicense" });
        }
      } catch {
        if (to.name !== "SettingLicense") {
          snackbar.showError("Failed to get your license info. Please check it and try again.");
          return next({ name: "SettingLicense" });
        }
      }
    }

    return next();
  },
);

export default router;
