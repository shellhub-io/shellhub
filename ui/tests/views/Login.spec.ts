import { createVuetify } from "vuetify";
import { flushPromises, mount, VueWrapper } from "@vue/test-utils";
import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";
import MockAdapter from "axios-mock-adapter";
import Login from "@/views/Login.vue";
import { usersApi } from "@/api/http";
import { store, key } from "@/store";
import { router } from "@/router";
import { envVariables } from "@/envVariables";
import { SnackbarPlugin } from "@/plugins/snackbar";

type LoginWrapper = VueWrapper<InstanceType<typeof Login>>;

describe("Login", () => {
  let wrapper: LoginWrapper;
  const vuetify = createVuetify();

  let mock: MockAdapter;

  beforeEach(() => {
    envVariables.isCloud = true;

    // Create a mock adapter for the usersApi instance
    mock = new MockAdapter(usersApi.getAxios());

    wrapper = mount(Login, {
      global: {
        plugins: [[store, key], vuetify, router, SnackbarPlugin],
      },
    });
  });

  afterEach(() => {
    vi.restoreAllMocks();
    mock.reset();
    wrapper.unmount();
  });

  it("Is a Vue instance", () => {
    expect(wrapper.vm).toBeTruthy();
  });

  it("Renders the component", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Renders the template with data", () => {
    expect(wrapper.find('[data-test="username-text"]').exists()).toBe(true);
    expect(wrapper.find('[data-test="password-text"]').exists()).toBe(true);
    expect(wrapper.find('[data-test="login-btn"]').exists()).toBe(true);
    expect(wrapper.find('[data-test="forgotPassword-card"]').exists()).toBe(true);
    expect(wrapper.find('[data-test="isCloud-card"]').exists()).toBe(true);
    expect(wrapper.find('[data-test="loadingToken-alert"]').exists()).toBe(false);
    expect(wrapper.find('[data-test="sso-btn"]').exists()).toBe(false);
    expect(wrapper.find('[data-test="or-divider-sso"]').exists()).toBe(false);
  });

  it("Renders enterprise only fragments", async () => {
    envVariables.isEnterprise = true;

    store.commit("users/setSystemInfo", {
      authentication: { local: true, saml: true },
    });

    await flushPromises();

    expect(wrapper.find('[data-test="sso-btn"]').exists()).toBe(true);
    expect(wrapper.find('[data-test="or-divider-sso"]').exists()).toBe(true);
  });

  it("disables fields and login button when envVariables.isEnterprise is true", async () => {
    envVariables.isCloud = false;
    envVariables.isEnterprise = true;

    store.commit("users/setSystemInfo", {
      authentication: { local: false, saml: true },
    });

    await flushPromises();

    const usernameField = wrapper.find('[data-test="username-text"]').attributes().class;
    const passwordField = wrapper.find('[data-test="password-text"]').attributes().class;
    const loginButton = wrapper.find('[data-test="login-btn"]').attributes().class;

    expect(usernameField).toContain("v-input--disabled");
    expect(passwordField).toContain("v-input--disabled");
    expect(loginButton).toContain("v-btn--disabled");
  });

  it("calls the login action when the form is submitted", async () => {
    const responseData = {
      token: "fake-token",
      user: "test",
      name: "Test",
      id: "1",
      tenant: "fake-tenant",
      role: "administrator",
      email: "test@test.com",
      mfa: false,
    };

    // mock error below
    mock.onPost("http://localhost:3000/api/login").reply(200, responseData);

    const loginSpy = vi.spyOn(store, "dispatch");
    const routerPushSpy = vi.spyOn(router, "push");

    await wrapper.findComponent('[data-test="username-text"]').setValue("test");
    await wrapper.findComponent('[data-test="password-text"]').setValue("password");
    await wrapper.findComponent('[data-test="form"]').trigger("submit");
    await flushPromises();

    // Assert the login action dispatch
    expect(loginSpy).toHaveBeenCalledWith("auth/login", {
      username: "test",
      password: "password",
    });

    expect(wrapper.findComponent(".v-alert").exists()).toBeFalsy();
    expect(routerPushSpy).toHaveBeenCalledWith({
      path: "/",
      query: {},
    });
  });

  it("calls the mfa action when the login form is submitted", async () => {
    const responseData = {
      token: "fake-token",
      user: "test",
      name: "Test",
      id: "1",
      tenant: "fake-tenant",
      role: "administrator",
      email: "test@test.com",
      mfa: true,
    };

    // mock error below
    mock.onPost("http://localhost:3000/api/login").reply(200, responseData);

    const loginSpy = vi.spyOn(store, "dispatch");
    const routerPushSpy = vi.spyOn(router, "push");

    await wrapper.findComponent('[data-test="username-text"]').setValue("testuser");
    await wrapper.findComponent('[data-test="password-text"]').setValue("password");
    await wrapper.findComponent('[data-test="form"]').trigger("submit");
    await flushPromises();

    // Assert the login action dispatch
    expect(loginSpy).toHaveBeenCalledWith("auth/login", {
      username: "testuser",
      password: "password",
    });

    expect(wrapper.findComponent(".v-alert").exists()).toBeFalsy();
    expect(routerPushSpy).toHaveBeenCalledWith({ name: "MfaLogin" });
  });

  it("shows an error message for a 401 response", async () => {
    const loginSpy = vi.spyOn(store, "dispatch");

    // mock error below
    mock.onPost("http://localhost:3000/api/login").reply(401);

    await wrapper.findComponent('[data-test="username-text"]').setValue("testuser");
    await wrapper.findComponent('[data-test="password-text"]').setValue("password");
    await wrapper.findComponent('[data-test="form"]').trigger("submit");
    await flushPromises();

    // Assert the login action dispatch
    expect(loginSpy).toHaveBeenCalledWith("auth/login", {
      username: "testuser",
      password: "password",
    });

    expect(wrapper.findComponent('[data-test="invalid-login-alert"]').exists());
  });

  it("redirects to ConfirmAccount route on 403 response", async () => {
    const loginSpy = vi.spyOn(store, "dispatch");
    const routerPushSpy = vi.spyOn(router, "push");

    // mock error below
    mock.onPost("http://localhost:3000/api/login").reply(403);

    await wrapper.findComponent('[data-test="username-text"]').setValue("testuser");
    await wrapper.findComponent('[data-test="password-text"]').setValue("password");
    await wrapper.findComponent('[data-test="form"]').trigger("submit");
    await flushPromises();

    // Assert the login action dispatch
    expect(loginSpy).toHaveBeenCalledWith("auth/login", {
      username: "testuser",
      password: "password",
    });

    expect(wrapper.findComponent(".v-alert").exists()).toBeFalsy();

    // Assert the redirection
    expect(routerPushSpy).toHaveBeenCalledWith({
      name: "ConfirmAccount",
      query: { username: "testuser" },
    });
  });

  it("locks account after 10 failed login attempts", async () => {
    const username = "testuser";
    const maxAttempts = 10;
    const lockoutDuration = 7 * 24 * 60 * 60; // 7 days in seconds
    let attempts = 0;

    mock.onPost("http://localhost:3000/api/login").reply((config) => {
      const { username: reqUsername, password } = JSON.parse(config.data);
      if (reqUsername === username && password === "wrongpassword") {
        attempts++;
        if (attempts >= maxAttempts) {
          return [429, {}, { "x-account-lockout": lockoutDuration.toString() }];
        }
        return [401];
      }
      return [200, { token: "fake-token" }];
    });

    // Simulate 10 failed login attempts
    for (let i = 0; i < maxAttempts; i++) {
      wrapper.findComponent('[data-test="username-text"]').setValue(username);
      wrapper.findComponent('[data-test="password-text"]').setValue("wrongpassword");
      wrapper.findComponent('[data-test="form"]').trigger("submit");
      // eslint-disable-next-line no-await-in-loop
      await flushPromises();
    }

    // Ensure the account is locked out
    expect(wrapper.findComponent('[data-test="invalid-login-alert"]').exists()).toBeTruthy();
  });
});
