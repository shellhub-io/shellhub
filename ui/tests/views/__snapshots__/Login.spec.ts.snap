// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login > Renders the component 1`] = `
"<div class="v-container v-locale--is-ltr">
  <!--v-if-->
  <!--v-if-->
  <transition-stub name="slide-y-reverse-transition" appear="false" persisted="false" css="true">
    <!---->
  </transition-stub>
  <transition-stub name="slide-y-reverse-transition" appear="false" persisted="false" css="true">
    <!---->
  </transition-stub>
  <form class="v-form" novalidate="" data-test="form">
    <div class="v-col">
      <div class="v-input v-input--horizontal v-input--center-affix v-input--density-default v-theme--light v-locale--is-ltr v-text-field" data-test="username-text">
        <!---->
        <div class="v-input__control">
          <div class="v-field v-field--center-affix v-field--prepended v-field--variant-filled v-theme--light v-locale--is-ltr">
            <div class="v-field__overlay"></div>
            <div class="v-field__loader">
              <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
                <!---->
                <div class="v-progress-linear__background bg-primary"></div>
                <div class="v-progress-linear__buffer bg-primary" style="width: 0%;"></div>
                <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                  <div class="v-progress-linear__indeterminate">
                    <div class="v-progress-linear__indeterminate long bg-primary"></div>
                    <div class="v-progress-linear__indeterminate short bg-primary"></div>
                  </div>
                </transition-stub>
                <!---->
              </div>
            </div>
            <div class="v-field__prepend-inner"><i class="mdi-account mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i>
              <!---->
            </div>
            <div class="v-field__field" data-no-activator=""><label class="v-label v-field-label v-field-label--floating" aria-hidden="true" for="input-v-0">
                <!---->Username or email address
              </label><label class="v-label v-field-label" for="input-v-0">
                <!---->Username or email address
              </label>
              <!----><input size="1" type="text" id="input-v-0" aria-describedby="input-v-0-messages" required="" class="v-field__input" value="">
              <!---->
            </div>
            <!---->
            <!---->
            <div class="v-field__outline">
              <!---->
              <!---->
            </div>
          </div>
        </div>
        <!---->
        <div id="input-v-0-messages" class="v-input__details" role="alert" aria-live="polite">
          <transition-group-stub name="slide-y-transition" tag="div" appear="false" persisted="false" css="true" class="v-messages">
            <!---->
          </transition-group-stub>
          <!---->
        </div>
      </div>
      <div class="v-input v-input--horizontal v-input--center-affix v-input--density-default v-theme--light v-locale--is-ltr v-text-field" data-test="password-text">
        <!---->
        <div class="v-input__control">
          <div class="v-field v-field--appended v-field--center-affix v-field--prepended v-field--variant-filled v-theme--light v-locale--is-ltr">
            <div class="v-field__overlay"></div>
            <div class="v-field__loader">
              <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
                <!---->
                <div class="v-progress-linear__background bg-primary"></div>
                <div class="v-progress-linear__buffer bg-primary" style="width: 0%;"></div>
                <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                  <div class="v-progress-linear__indeterminate">
                    <div class="v-progress-linear__indeterminate long bg-primary"></div>
                    <div class="v-progress-linear__indeterminate short bg-primary"></div>
                  </div>
                </transition-stub>
                <!---->
              </div>
            </div>
            <div class="v-field__prepend-inner"><i class="mdi-lock mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i>
              <!---->
            </div>
            <div class="v-field__field" data-no-activator=""><label class="v-label v-field-label v-field-label--floating" aria-hidden="true" for="input-v-2">
                <!---->Password
              </label><label class="v-label v-field-label" for="input-v-2">
                <!---->Password
              </label>
              <!----><input size="1" type="password" id="input-v-2" aria-describedby="input-v-2-messages" required="" class="v-field__input" value="">
              <!---->
            </div>
            <!---->
            <div class="v-field__append-inner">
              <!----><i class="mdi-eye-off mdi v-icon notranslate v-theme--light v-icon--size-default v-icon--clickable" role="button" aria-hidden="false" tabindex="0" aria-label="Password appended action"></i>
            </div>
            <div class="v-field__outline">
              <!---->
              <!---->
            </div>
          </div>
        </div>
        <!---->
        <div id="input-v-2-messages" class="v-input__details" role="alert" aria-live="polite">
          <transition-group-stub name="slide-y-transition" tag="div" appear="false" persisted="false" css="true" class="v-messages">
            <!---->
          </transition-group-stub>
          <!---->
        </div>
      </div>
      <div class="v-card-actions justify-center pa-0"><button type="submit" class="v-btn v-btn--block v-btn--disabled v-btn--slim v-theme--light text-primary v-btn--density-default v-btn--size-default v-btn--variant-tonal" disabled="" data-test="login-btn"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
          <!----><span class="v-btn__content" data-no-activator=""> Login </span>
          <!---->
          <!---->
        </button></div>
    </div>
  </form>
  <div class="v-col">
    <div class="v-card-subtitle d-flex align-center justify-center pa-4 mx-auto pt-0 pb-0" data-test="forgotPassword-card"> Did you <a href="/forgot-pass" class="ml-1"> Forgot your Password? </a></div>
    <div class="v-card-subtitle d-flex align-center justify-center pa-4 mx-auto" data-test="isCloud-card"> Don't have an account? <a href="/sign-up" class="ml-1"> Sign up here </a></div>
  </div>
  <!--v-if-->
</div>"
`;
